src/hilok.cpp

```diff
@@ -213,41 +219,47 @@ void HiLok::rename(std::string_view path_from, std::string_view path_to, bool bl
             if (!cur_to->m_mut.unsafe_clone_lock_shared(leaf_from_node->m_mut, block, secs)) {
                 throw HiErr("unable to lock rename dest");
             }
-        } else {
-            cur_to.reset(); // refcount for erase
+        }
+    }
 
-            while (it_from != it_from.end()) {
-                // uncommon ancestor of source must be released
-                ++it_from;
+    std::vector<std::shared_ptr<HiKeyNode>> to_erase;
 
-                auto it = m_map.find(from_key);
+    while (it_from != it_from.end()) {
+        // uncommon ancestor of source must be released
 
-                // we already tested for this above, and we have a mutex, should never happen
-                assert(it != m_map.end());
+        auto it = m_map.find(from_key);
 
-                cur_from = it->second;
+        // we already tested for this above, and we have a mutex, should never happen
+        assert(it != m_map.end());
+
+        cur_from = it->second;
 
 #ifdef HILOK_TRACE
-                std::cout << "clon un: " << from_key.first << "/" << from_key.second << ":" << cur_from << std::endl;
+        std::cout << "clon un: " << from_key.first << "/" << from_key.second << ":" << cur_from << std::endl;
 #endif
-                // unlock uncommon ancestors of the source
-                cur_from->m_mut.unsafe_clone_unlock_shared(leaf_from_node->m_mut);
+        // unlock uncommon ancestors of the source
+        cur_from->m_mut.unsafe_clone_unlock_shared(leaf_from_node->m_mut);
 
-                from_key.first.reset(); // refcount for erase
-                // could have went to 0
-                erase_unsafe(cur_from);
-            
-                from_key = {cur_from, *it_from};
-            }
+        to_erase.push_back(cur_from);
 
-            // keep leaf locks, change key
-            m_map.erase(leaf_from_node->m_key);
-            leaf_from_node->m_key = key;
-            m_map[key] = leaf_from_node;        
-       } 
+        from_key = {cur_from, *it_from};
+        
+        ++it_from;
+    }
+   
+    // lower refs for erase hint
+    from_key.first.reset();
+    cur_to.reset();
+    cur_from.reset();
+
+    for (auto nod : to_erase) {
```

Comment:
This will make a new `shared_ptr`, incrementing refcount by one. Any reason not to do

```suggestion
    for (auto &nod : to_erase) {
```

instead?

Reply:
thanks